### DISCUSSION TITLE
Provide an alternate communication channel via gRPC

### DISCUSSION THREAD

**Discussion by gargakshit** [meilisearch/product#143] [Feedback & Feature Proposal]
**Is your feature request related to a problem? Please describe.**
A lot of languages don't have stable client libraries. `gRPC` would allow to codegen clients for any language that has a protobuf and gRPC package. Also gRPC will have a bit less overhead as gRPC's HTTP/2 connections are long lived, saving the overhead for re-creating an HTTP connection for each request. A gRPC based communication channel would be super handy.

**Describe the solution you'd like**
Implement a gRPC based communication channel to do operations and queries.

**Comment by gmourier** [comment]
Hi @gargakshit! Thanks for the feedback. We thought about it a while ago but not many people are asking for it yet. Looking at the [roadmap for Q3/Q4](https://github.com/meilisearch/product/discussions/61), it's something we could consider during 2022.

Do you see other advantages to have gRPC? I see an interest to identify breaking changes via .proto files contracts.

Thanks!

**Comment by gargakshit** [comment]
> Do you see other advantages to have gRPC? I see an interest to identify breaking changes via .proto files contracts.

Other advantages include
- Less overhead than http/1.1 as http/2 is a binary protocol, and gRPC connections are long lived preventing the overhead of re-connecting on every request. HTTP's `Keep-Alive` does solve this to some extent, but gRPC would still have less overhead
- Generating clients for languages that don't already have a package available. This would be useful as languages like `elixir` don't have a mature and maintained meilisearch client, while they do have a mature gRPC client/server
- Maybe streaming changes? I don't know how feasible or useful this would be in the context of a search engine, but gRPC has unary and bidirectional streaming

**Comment by ahmednfwela** [comment]
I am in favor of this, since this will finally prevent the need for `waitForTask` polling mechanisms.
it will also pave the road for listening to a document for changes (and maybe even a query, like firebase)
